Gradual typing

akaihola opened the pull request:
Here's another stab at adding typing since #639 was closed.

I believe this is the minimal Mypy configuration and minimal set of changes possible to make Mypy pass, have it check _some_ types, and allow development to continue without requiring full type hints everywhere.

Most notably, functions with no type hints are not checked by Mypy at all with this configuration. This allows adding type hints just only to select sections of the code base. It is still of course possible to gradually add typing overall and increase Mypy's strictness if desired.

See [Using mypy with an existing codebase](https://mypy.readthedocs.io/en/stable/existing_code.html) for more information.

ruancomelli commented on aider/coders/base_prompts.py:
```diff
@@ -35,7 +38,7 @@ class CoderPrompts:
     )
 
-    repo_content_prefix = """Here are summaries of some files present in my git repository.
+    repo_content_prefix: str | None = """Here are summaries of some files present in my git repository.
```
The type union syntax `X | Y` is only available in Python 3.10, but [Aider still supports Python 3.9](https://github.com/paul-gauthier/aider/blob/e6037140be0105ee2575761b69c0e5ec7acf946b/pyproject.toml#L23).
Instead, this should be
```suggestion
    repo_content_prefix: typing.Optional[str] = """Here are summaries of some files present in my git repository.
```

ryanpeach replied:
Need to put that in the tests too it looks like:

`TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'`

nedtwigg commented:
I would think that typing an untyped python codebase is exactly the kind of thing that Aider itself would be great at. Typing the edges is trivial, that leads to type errors, fix those, that leads to new type errors, fix those, etc.

I also think if it's worth adding type annotations at all, then it's worth bumping to a Python version that supports the latest / greatest versions of them. Aider seems like more of an end-user thing than infrastructure library, seems okay for it to be more aggressive in bumping requirements than a typical library.

Faolain commented:
Amazing work as this will make aider improving itself even better (from experience llms do quite well with typed code)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Aider-AI/aider?pullRequest=1217) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Aider-AI/aider?pullRequest=1217) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Aider-AI/aider?pullRequest=1217) it.</sub>

akaihola commented:
I rebased on `main` and fixed all typing errors which had appeared meanwhile. Also added a requirements file for running Mypy. It includes Mypy itself and some type stub packages. The workflow now uses this requirements file as well as other requirements files needed for type checking.

ryanpeach commented:
Have you considered pyright instead? Pyright is fast enough to work in your vscode and neovim live, and can go in your pre-commit.

Also we should put typing into the github actions

ryanpeach commented on .github/workflows/lint.yml:
```diff
@@ -0,0 +1,43 @@
+      run: |
+        python -m pip install --upgrade pip
+        pip install '.[browser,dev,playwright,mypy]'
+
```
Add flake8 from the pre-commit and a black check

ryanpeach commented on .github/workflows/lint.yml:
```diff
@@ -0,0 +1,43 @@
+      run: |
+        python -m pip install --upgrade pip
+        pip install '.[browser,dev,playwright,mypy]'
+
```
add flake8 and black checks as they are in the pre-commit

ryanpeach replied:
Better to just add https://github.com/pre-commit/action

akaihola replied:
I'd like to keep this PR minimal and just about adding minimal typing and checks for types. So other linting workflows would be better added in separate PRs.

ryanpeach reviewed (approved):
Very minimal IMO. I like it (though I have no authority)

ryanpeach commented on .github/workflows/lint.yml:
```diff
@@ -0,0 +1,43 @@
+name: Linting
+
+on:
+  push:
```
Not sure why we need this on push

ryanpeach commented on .github/workflows/lint.yml:
```diff
@@ -0,0 +1,43 @@
+    - name: Set up Python 3.12
+      uses: actions/setup-python@v5
+      with:
+        python-version: "3.12"
```
Use a matrix like in the other workflows

ryanpeach reviewed (changes requested):
A few more lint comments

ryanpeach commented:
```
ERROR: Cannot install None, aider-chat and aider-chat[browser,dev,mypy,playwright]==0.1.dev1+g1aa8de2 because these package versions have conflicting dependencies.

The conflict is caused by:
    aider-chat 0.1.dev1+g1aa8de2 depends on attrs==24.3.0
    aiohttp 3.11.11 depends on attrs>=17.3.0
    jsonschema 4.23.0 depends on attrs>=22.2.0
    referencing 0.36.0 depends on attrs>=22.2.0
    aider-chat[browser,dev,mypy,playwright] 0.1.dev1+g1aa8de2 depends on attrs==24.3.0
    aider-chat[browser,dev,mypy,playwright] 0.1.dev1+g1aa8de2 depends on attrs==24.3.0; extra == "browser"
    aider-chat[browser,dev,mypy,playwright] 0.1.dev1+g1aa8de2 depends on attrs==24.2.0; extra == "mypy"

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip to attempt to solve the dependency conflict
```

https://github.com/OnScale/aider-stateless/actions/runs/12914071198/job/36012781606?pr=7

Faolain commented:
Completely agree here @paul-gauthier what are your thoughts?
